// customers/bin/Scripts/src/components/data-table/data-table.js
import RegisterComponent from "../../componentRegister";
import template from './data-table-template';

RegisterComponent('data-table', {
    script: {
        create: (props) => {
            return DataTable({ id: props.id, columnsDefinition: props.columns, rows: props.rows });
        },
        onCreated: (script) => script.initialize()
    },
    template: template
});

export default function DataTable({ id, columnsDefinition, rows }) {
    let columns = ko.observableArray([]);

    function initialize() {
        buildColumns();
    }

    function whenRowRendered(element) {
        if (element[1].rowIndex != rows().length) return;
        create();
    }

    function create() {
        return $('#' + id).DataTable({
            retrieve: true,
            ordering: true,
            searching: false,
            paging: false,
            info: true
        });
    }

    function buildColumns() {
        ko.utils.arrayPushAll(columns, _.map(columnsDefinition, (colDefinition) => buildColumn(colDefinition)));
    }

    function buildColumn(colDefinition) {
        return {
            name: colDefinition.name,
            label: colDefinition.label,
            type: colDefinition.type || 'text',
            align: colDefinition.align || 'left'
        };
    }

    return {
        initialize,
        tableId: id,
        columns,
        rows,
        whenRowRendered
    };
}
