// customers/bin/Scripts/src/components/search-select-remote/search-select-remote.js
import ComponentRegister from '../../componentRegister';
import template from './search-select-remote-template';

ComponentRegister('search-select-remote', {
    script: {
        create: (props) => {
            const params = {
                label: props.label,
                url: props.url,
                placeholder: props.placeholder,
                enableEvents: props.enableEvents,
                allowUnselect: props.allowUnselect,
                options: props.options,
                selectedFromCaller: props.selected,
                optionsText: props.optionsText,
                enabledFromCaller: props.enabled,
                whenSelected: props.whenSelected
            };
            return new SearchSelectRemote(params);
        },
        onCreated: (script) => script.initialize()
    },
    template: template
});

export default function SearchSelectRemote(
    { label, placeholder, selectedFromCaller, enabledFromCaller, options, enableEvents, allowUnselect, url, optionsText, whenSelected }) {
    allowUnselect = (allowUnselect != undefined) ? allowUnselect : true;
    optionsText = optionsText || 'description';

    const enabled = ko.computed(() => isComponentEnabled());
    const eventsEnabled = areEventsEnabled();
    const minInputLength = 3;
    const selected = ko.observable();

    function initialize() {
        if (selectedFromCaller()) {
            selected(selectedFromCaller());
        }
    }

    function isComponentEnabled() {
        if (enabledFromCaller) return enabledFromCaller();
        return true;
    }

    function areEventsEnabled() {
        if (enableEvents !== undefined) {
            return ko.unwrap(enableEvents);
        }
        return true;
    }

    selected.subscribe((value) => {
        if (!value) return;
        if (selectedFromCaller() && value.id == selectedFromCaller().id) return;
        if (eventsEnabled && whenSelected) {
            selectedFromCaller(value);
            whenSelected(value);
        }
    });

    selectedFromCaller.subscribe((value) => {
        if (!value) return;
        if (selected() && selected().id == value.id) return;
        selected(value);
    });

    return {
        initialize,
        label,
        placeholder,
        enabled,
        options,
        selected,
        url,
        optionsText,
        allowUnselect,
        minInputLength
    };
}
