// customers/obj/Release/Package/PackageTmp/Scripts/src/domain/configurator/configurator-template.js
/* eslint-disable */// Autogenerated with Gulp (task: parse-components-templates)
// Dont change this file, change the .html file instead and run Gulp
export default `
<div class="box box-default">
    <div class="box-header with-border">
        <h3 class="box-title" data-bind="text: labels.title"></h3>
    </div>
    <div class="box-body">
        <div class="form-horizontal">
            <model-selector params="modelSelected: modelSelected,
                                    whenModelSelected: whenModelSelected, 
                                    whenModelUnselected: whenModelUnselected, 
                                    whenModelGroupSelected: whenModelGroupSelected">
            </model-selector>
            <color-selector params="options: colorOptions, selected: selectedColor"></color-selector>
            <interior-selector params="options: interiorOptions, selected: selectedInterior"></interior-selector>
        </div>
    </div>
    <spinner params="show: loadingCatalog"></spinner>
</div>
<equipments-selector params="equipments: availableEquipments, 
                             whenEquipmentSelected: whenEquipmentSelected,
                             whenEquipmentUnselected: whenEquipmentUnselected,
                             loadingFlag: loadingCatalog">
</equipments-selector>
<combination-selector params="visible: combinationSelectorVisible, options: combinationOptions, isSelectionMandatory: isCombinationMandatory, whenSelected: whenCombinationSelected"></combination-selector>`;
