// customers/obj/Release/Package/PackageTmp/bin/Scripts/src/components/search-select/search-select.js
import ComponentRegister from '../../componentRegister';
import template from './search-select-template';

ComponentRegister('search-select', {
    script: {
        create: (props) => {
            const params = {
                label: props.label,
                placeholder: props.placeholder,
                options: props.options,
                enableEvents: props.enableEvents,
                allowUnselect: (props.allowUnselect != undefined) ? props.allowUnselect : true,
                selectedFromCaller: props.selected,
                whenSelected: props.whenSelected,
                urlToSearch: props.urlToSearch,
                optionsText: props.optionsText || 'description'
            };
            return new SearchSelect(params);
        },
        onCreated: (script) => script.initialize()
    },
    template: template
});

export default function SearchSelect({ label, placeholder, options, selectedFromCaller, enableEvents, allowUnselect, whenSelected, urlToSearch, optionsText }) {
    const enabled = ko.computed(() => options != undefined && options().length > 0);
    const eventsEnabled = areEventsEnabled();
    const minInputLength = 3;
    const selected = ko.observable();

    function initialize() {
        if (selectedFromCaller()) {
            selected(selectedFromCaller());
        }
    }

    function areEventsEnabled() {
        if (enableEvents !== undefined) {
            return ko.unwrap(enableEvents);
        }
        return true;
    }

    selected.subscribe(() => {
        if (selected() == selectedFromCaller()) return;
        if (eventsEnabled && whenSelected) {
            selectedFromCaller(selected());
            whenSelected(selected());
        }
    });

    selectedFromCaller.subscribe(() => {
        if (selected() == selectedFromCaller()) return;
        selected(selectedFromCaller());
    });

    return {
        initialize,
        label,
        placeholder,
        options,
        enabled,
        selected,
        allowUnselect,
        urlToSearch,
        minInputLength,
        optionsText
    };
}
