// customers/obj/Release/Package/PackageTmp/bin/Scripts/src/DialogPublisher.js
export let DialogPublisher = (function(){

    let onSuccessCallback = () => {};
    let onWarningCallback = () => {};
    let onErrorCallback = () => {};
    let onConfirmationCallback = () => {};
    let onQuestionCallback = () => {};

    function success({message, htmlTemplate}) {
        onSuccessCallback({message, htmlTemplate});
    }

    function onSuccess(callback){
        onSuccessCallback = callback;
    }

    function onWarning(callback){
        onWarningCallback = callback;
    }

    function warning({title, message, htmlTemplate}){
        onWarningCallback({title, message, htmlTemplate});
    }

    function onError(callback){
        onErrorCallback = callback;
    }

    function error({title, message, htmlTemplate}){
        onErrorCallback({title, message, htmlTemplate});
    }

    function onConfirmation(callback) {
        onConfirmationCallback = callback;
    }

    function confirmation({title, message, callback}) {
        onConfirmationCallback({title, message, callback});
    }

    function onQuestion(callback) {
        onQuestionCallback = callback;
    }

    function ask({title, message, yesButton, noButton, yesCallback, noCallback}) {
        onQuestionCallback({title, message, yesButton, noButton, yesCallback, noCallback});
    }

    return {
        success,
        onSuccess,
        warning,
        onWarning,
        error,
        onError,
        onConfirmation,
        confirmation,
        onQuestion,
        ask
    };
})();
