// customers/Scripts/bin/aida/aida-spec-core.js
require=(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports["default"] = spyCallback;

function spyCallback() {
    var called = false;
    var switcher = function switcher() {
        return called;
    };
    var f = function f() {
        called = true;
    };
    var fWrapper = function fWrapper() {
        f();
    };
    fWrapper.isCalled = switcher;

    return fWrapper;
}

module.exports = exports["default"];

},{}],2:[function(require,module,exports){
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports["default"] = stubTranslationService;

function stubTranslationService() {
    return {
        translate: function translate(key) {
            return key;
        }
    };
}

module.exports = exports["default"];

},{}],3:[function(require,module,exports){
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.exportApi = exportApi;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var _stubTranslationService = require("./stubTranslationService");

var _stubTranslationService2 = _interopRequireDefault(_stubTranslationService);

var _spyCallback = require("./spyCallback");

var _spyCallback2 = _interopRequireDefault(_spyCallback);

function exportApi() {
    return {
        StubTranslationService: _stubTranslationService2["default"],
        SpyCallback: _spyCallback2["default"]
    };
}

var aidaSpecCore = require('aida-spec-core').exportApi();
window.aidaSpecCore = aidaSpecCore;

},{"./spyCallback":4,"./stubTranslationService":5,"aida-spec-core":"aida-spec-core"}],4:[function(require,module,exports){
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports["default"] = spyCallback;

function spyCallback() {
    var called = false;
    var switcher = function switcher() {
        return called;
    };
    var f = function f() {
        called = true;
    };
    var fWrapper = function fWrapper() {
        f();
    };
    fWrapper.isCalled = switcher;

    return fWrapper;
}

module.exports = exports["default"];

},{}],5:[function(require,module,exports){
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports["default"] = stubTranslationService;

function stubTranslationService() {
    return {
        translate: function translate(key) {
            return key;
        }
    };
}

module.exports = exports["default"];

},{}],"aida-spec-core":[function(require,module,exports){
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.exportApi = exportApi;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var _stubTranslationService = require("./stubTranslationService");

var _stubTranslationService2 = _interopRequireDefault(_stubTranslationService);

var _spyCallback = require("./spyCallback");

var _spyCallback2 = _interopRequireDefault(_spyCallback);

function exportApi() {
    return {
        StubTranslationService: _stubTranslationService2["default"],
        SpyCallback: _spyCallback2["default"]
    };
}

var aidaSpecCore = require('aida-spec-core').exportApi();
window.aidaSpecCore = aidaSpecCore;

},{"./spyCallback":1,"./stubTranslationService":2,"aida-spec-core":"aida-spec-core"}]},{},[3,4,5])
