// customers/bin/Scripts/src/restClient.js
export default function RestClient(apiBasePath) {

    function get(resourceUrl, callback, errorCallback) {
        $.ajax({
            url: apiBasePath + resourceUrl,
            headers: {
                "SIMA-Session-Token": "0"
            },
            success: callback,
            error: errorCallback
        });
    }

    function post(resourceUrl, headers, data, callback, errorCallback) {
        $.ajax({
            url: apiBasePath + resourceUrl,
            type: 'POST',
            headers: headers,
            dataType: "json",
            contentType: "application/json; charset=utf-8",
            data: JSON.stringify(data),
            success: (result, status, xhr) => {
                let location = xhr.getResponseHeader('Location');
                callback(result, location);
            },
            error: errorCallback
        });
    }

    function remove(url, headers, callback, errorCallback) {
        $.ajax({
            url: apiBasePath + url,
            type: 'DELETE',
            headers: headers,
            success: callback,
            error: errorCallback
        });
    }

    function put(url, headers, data, callback, errorCallback) {
        $.ajax({
            url: apiBasePath + url,
            type: 'PUT',
            headers: headers,
            dataType: "json",
            contentType: "application/json; charset=utf-8",
            data: JSON.stringify(data),
            success: callback,
            error: errorCallback
        });
    }

    return { get, post, put, remove };
}
