// customers/obj/Release/Package/PackageTmp/Scripts/spec/domain/catalogHelper.js
let CatalogHelper = (function () {
    const airConditioner = {
        code: "PH1",
        description: "Aire acondicionado Climatic",
        esmEquipment: true,
        category: "Paquete"
    };
    const radio = {
        code: "RFZ",
        description: "Radio CD 215 con función MP3 y conexión Aux-In",
        esmEquipment: true,
        category: "Paquete"
    };
    const camera = {
        code: "PR1",
        category: "Comm",
        description: "Camera"
    };
    const bluetooth = {
        code: "PR2",
        category: "Comm",
        description: "Bluetooth"
    };
    const navigationPack = {
        code: "77G",
        description: "Datos de navegación para Europa en memoria interna",
        esmEquipment: false,
        type: "Equipment",
        category: "Comm"
    };
    const safetyPackage = {
        code: "PSA",
        description: "up & safety package",
        esmEquipment: false,
        category: "Paquete",
        type: "Equipment",
        equipments: []
    };
    const equipmentWhitoutCategory  = {
        code: "PH2",
        description: "Climatizador",
        esmEquipment: true,
        category: undefined,
        family: "FAM"
    };    

    const red = { description: "red", code: "X1X1" };
    const blue = { description: "blue", code: "X2X2" };
    const black = { description: "black", code: "A1A1" };

    const leather = { description: "cuero", code: "A1" };
    const sky = { description: "sky", code: "A2" };
    const corduroy = {
        code: "A3",
        hasForcedCombinations: true,
        compulsoryCombinations: [[camera], [bluetooth]]
    };


    const catalog = {
        equipments: [ airConditioner, radio, navigationPack, camera, bluetooth ],
        packages: [safetyPackage],
        colors: [red, blue, black],
        interiors: [leather, sky, corduroy],
        modelCompulsoryCombinationChoices: [],
        relationship: [
            { colorCode: red.code, interiorCode: leather.code },
            { colorCode: blue.code, interiorCode: sky.code }
        ]
    };

    return {
        camera,
        bluetooth,
        navigationPack,
        airConditioner, 
        radio,
        red,
        blue,
        black,
        leather,
        sky,
        corduroy,
        safetyPackage,
        equipmentWhitoutCategory,
        catalog
    };
})();

export default CatalogHelper;
