// customers/Scripts/src/domain/customer-detail-section/customer-detail-section-template.js
/* eslint-disable */// Autogenerated with Gulp (task: parse-components-templates)
// Dont change this file, change the .html file instead and run Gulp
export default `
<section-box params="{ title: translator.translate('SharedDomain.FormData'), isShown: isShown }">
    <section-body>
        <customer-type-selector params="{
            isCompany: props.isCompany,
            translator: translator }">
        </customer-type-selector>
        <customer-form params="{
            id: props.id,
            name: props.name,
            firstSurname: props.firstSurname,
            secondSurname: props.secondSurname,
            birthDate: props.birthDate,
            gender: props.gender,
            taxId: props.taxId,
            phoneNumber: props.phoneNumber,
            email: props.email,
            company: props.company,
            address: props.address,
            postalCode: props.postalCode,
            city: props.city,
            state: props.state,
            country: props.country,

            isEdition: props.isEdition,
            availableCountries: props.availableCountries,
            genderOptions: props.genderOptions,
            isCompany: props.isCompany,
            translator: translator,
            isNameMissing: props.isNameMissing,
            isMissing: props.isMissing }">
        </customer-form>
    </section-body>     
    <section-footer>
        <button class="btn btn-primary" data-bind="click: props.saveCustomer, enable: props.canSaveCustomer, text: translator.translate('SharedDomain.SaveButton.Text')"></button>
        <button class="btn btn-default" data-bind="click: props.hide, visible: props.showCancel, text: translator.translate('SharedDomain.CancelButton.Text')"></button>
    </section-footer>
    <!--<div data-backdrop="static" data-keyboard="false" data-bind="modal: {
                    visible: props.customerSelector.visible,
                    header: { name: 'header_without_close', data: { label: translator.translate('VehicleOrders.CustomerSelector.Header.Label') }},
                    body: { name: 'customers_selector', data: props.customerSelector },
                    footer: { name: 'simple_footer', data: { closeVisible: props.customerSelector.mandatorySelection, closeLabel: translator.translate('SharedDomain.Neither.Action') }}
                }">
    </div>
    <script type="text/html" id="header_without_close">
        <h3 data-bind="text: label"></h3>  
    </script>
    <script type="text/html" id="customers_selector">
        <div data-bind="foreach: selectableCustomers">
            <div class="selectablebox" data-bind="click: select" style="margin: 5px">
                <span data-bind="text: customer.id + ' - '"></span>
                <span data-bind="text: customer.name"></span>
                <span data-bind="text: customer.firstSurname"></span>
                <span data-bind="text: customer.secondSurname"></span>
                <span> - </span>
                <span data-bind="text: customer.taxId"></span>
            </div>
        </div>
    </script>-->
    <script type="text/html" id="simple_footer">
        <!-- ko if: !$data.closeVisible() -->
        <a href="#" class="btn btn-default" data-bind="html: closeLabel" data-dismiss="modal"></a>
        <!-- /ko -->
    </script>
    <!-- /ko -->
</section-box>`;
